src/xwayland/xserver.rs

```diff
@@ -1,509 +1,378 @@
+        // SAFETY: we never mutate display_fd, and it lives as long as the instance.
+        unsafe { BorrowedFd::borrow_raw(guard.display_fd.as_raw_fd()) }
+    }
 
-    // all is ready, we can do the fork dance
-    let child_stdout = match spawn_xwayland(display, wl_x11, x_wm_x11, &x_fds, envs) {
-        Ok(child_stdout) => child_stdout,
-        Err(e) => {
-            error!(error = ?e, "XWayland failed to spawn");
-            return Err(e);
-        }
-    };
-
-    let loop_inner = inner.clone();
-    loop_handle
-        .insert_source(
-            Generic::<_, io::Error>::new(child_stdout, Interest::READ, Mode::Level),
-            move |_, child_stdout, _| {
-                // the closure must be called exactly one time, this cannot panic
-                // Safety: xwayland_ready will not close the child_stdout
-                xwayland_ready(&loop_inner, unsafe { child_stdout.get_mut() });
-                Ok(calloop::PostAction::Remove)
-            },
-        )
-        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
-
-    let client_fd = wl_me.as_raw_fd();
-
-    let data_map = UserDataMap::new();
-    user_data(&data_map);
-    let client = dh.insert_client(
-        wl_me,
-        Arc::new(XWaylandClientData {
-            inner: inner.clone(),
-            #[cfg(feature = "wayland_frontend")]
-            compositor_state: CompositorClientState::default(),
-            data_map,
-        }),
-    )?;
-    guard.instance = Some(XWaylandInstance {
-        display_lock: lock,
-        wayland_client: client,
-        wayland_client_fd: client_fd,
-        wm_fd: Some(x_wm_me),
-    });
-
-    Ok(display)
+    /// Checks if the XWayland instance is ready. If XWayland has fully started,
+    /// this will return the X11 socket connected to the running instance.
+    ///
+    /// Calling `take_socket` successfully transfers ownership of the connection
+    /// to the caller. After returning `Some` the first time, it will always
+    /// return `None`.
+    ///
+    /// An `Err` result is only returned if the XWayland instance has exited
+    /// unexpectedly.
+    ///
+    /// This is a low-level method. Using the instance as an event source is the
+    /// recommended way to interact with it.
+    pub fn take_socket(&mut self) -> std::io::Result<Option<UnixStream>> {
+        self.inner.lock().unwrap().take_socket()
+    }
 }
 
-/// An event source for monitoring XWayland status
-///
-/// You need to insert it in a [`calloop`] event loop to handle the events it produces,
-/// of type [`XWaylandEvent`], which notify you about startup and shutdown of the Xwayland
-/// instance.
 #[derive(Debug)]
-#[must_use]
-pub struct XWaylandSource {
-    channel: Channel<XWaylandEvent>,
+struct Instance {
+    display_lock: X11Lock,
+    x11_socket: Option<UnixStream>,
+    display_fd: OwnedFd,
 }
 
-impl calloop::EventSource for XWaylandSource {
+impl calloop::EventSource for XWayland {
```

Comment:
The display number is technically inaccessible unless you use a Dispatcher since LoopHandle::insert_source needs to own the event source. I think a way around that could be providing XWayland via the metadata associated type but that may be an issue with double borrows.

Reply:
The event has the display number on it, or you can fetch it before you `insert_source`.